### PR TITLE
Add basic support for sass/scss files to Template manager editing

### PR DESCRIPTION
**Pull request by brianteeman**
#### Summary of Changes

Add ability to see sass files in the list of available files in the tree
Add ability to create a new file of type sass/scss
Add ability to open, edit and save a sass/scss file
#### Testing Instructions

Install patch and go to template manager and select a template (not a style)
Select New File from the toolbar
Create a new file and select sass from the file type dropdown
Check that file can be seen in the folder tree and can be selected, edited, saved and deleted

### CODE REVIEW

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on cb02eb7c12726a2d8e33ae5235b648b219d415ae<br /><br />File cannot be seen, but it exist. Cause make a new File with same Name and Place in Folder got Error 

"File with the same name already exists. Failed to create file."

Test on Joomla! 3.5.2-dev<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on cb02eb7c12726a2d8e33ae5235b648b219d415ae<br /><br />Works fine, but i need to many added 'sass' extension to templates "Options" to view the created file in the list os files.

After that i did:
1. Selected css dir
2. New file type 'sass' named 'template'
3. Created and can view and edit it.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by andrepereiradasilva** [comment]
BTW you can also have `scss` file extensions? right?

**Comment by brianteeman** [comment]
@andrepereiradasilva yes you can but its not as common

I think I see the problem you both had and I dont know how to fix it. I am adding the file type sass in the options but its not seen until you open and save the options<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by andrepereiradasilva** [comment]
> @andrepereiradasilva yes you can but its not as common

Yes, but it seems in the new version, so if adding sass, shouldn't we also add scss?

> I think I see the problem you both had and I dont know how to fix it. I am adding the file type sass in the options but its not seen until you open and save the options

i only see two ways: force updating the options parameters in the database on joomla update (IMO not B/C)... or adding a post install message to inform users of the change and what they need to do. I don't like neither. Most users don't even use this options. Is it worth it?

**Comment by brianteeman** [comment]
We could add scss no problem

I think the database update is what has been done before but I am not sure.
Maybe someone else knows. I definitely wouldnt go with the post-install
message

On 12 April 2016 at 17:29, andrepereiradasilva <notifications@github.com>
wrote:

> @andrepereiradasilva <https://github.com/andrepereiradasilva> yes you can
> but its not as common
>
> Yes, but it seems in the new version, so if adding sass, shouldn't we also
> add scss?
>
> I think I see the problem you both had and I dont know how to fix it. I am
> adding the file type sass in the options but its not seen until you open
> and save the options
>
> i only see two ways: force updating the options parameters in the database
> on joomla update (IMO not B/C)... or adding a post install message to
> inform users of the change and what they need to do. I don't like neither.
> Most users don't even use this options. Is it worth it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9871#issuecomment-208992190>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by andrepereiradasilva** [comment]
> We could add scss no problem

yes

**Comment by brianteeman** [comment]
Updated to include scss files<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 5d83f8baa6d75787e2ed2c59a2e0e2bce1091cfb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by zero-24** [comment]
@brianteeman i don't think we should overrwrite it with a db update as this can ause problems if somone has add / cahnge this setting bevor the update. (e.g. removed some extensions) if we include the db update we revert all changes he made.

So this one will be only for new installs.

The other way arround could be to analyse this setting during the update (script.php) and only add this two values in (if not set bevor) but leave the other one allown.

But i think this is to mich code for a that simple change and if someone is come up and wantbto use it there can be a document page that explains the place to add the extensions.

Another way arround could be a messagebif someone try to create a file with a extesion that is not in the allowed extension setting amd told the user "please add the extension to the settings" or something like that.

**Comment by zero-24** [comment]
The seccond option would also catch the case that somone manually removend some extensions form that list.

**Comment by brianteeman** [comment]
I agree with your assessment - any idea how to do the second option?

On 12 April 2016 at 18:13, zero-24 <notifications@github.com> wrote:

> The seccond option would also catch the case that somone manually removend
> some extensions form that list.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9871#issuecomment-209012032>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by zero-24** [comment]
Please have a look at: https://github.com/brianteeman/joomla-cms/pull/9

**Comment by brianteeman** [comment]
I have merged the suggestion from @zero-24 please can you all test again<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

File: administrator/components/com_templates/models/template.php [hunk 1]
```diff
@@ -524,9 +518,6 @@ public function save($data)
 			return false;
 		}
 
-		$explodeArray = explode('.', $fileName);
```
Comment (wilsonge) [inline administrator/components/com_templates/models/template.php]: What was this for and why are we removing it?

Comment (wilsonge) [inline administrator/components/com_templates/models/template.php]: OK I'm pretty sure this is going to break the less stuff as $ext is no longer defined

Comment (brianteeman) [inline administrator/components/com_templates/models/template.php]: This was from @zero-24 commit to my repo. Didnt understand it but took him at his word

Comment (wilsonge) [inline administrator/components/com_templates/models/template.php]: I think this looks fine with the exception of the removal of these two lines. Was there a reason for it @zero-24 ? Or just a mistake?

Comment (zero-24) [inline administrator/components/com_templates/models/template.php]: Thanks @wilsonge you are correct i have changed this part many times on coding but i miss the final code review see here: https://github.com/brianteeman/joomla-cms/pull/10

**Comment by brianteeman** [comment]
I have merged @zero-24 latest correction

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 25165893ec8ef1e474d06feb5da501432afd9c3e<br /><br />Tested to create sass/scss-Files on Beez3, Hathor, Isis and Protostar. Got now Message why Files aren't shown.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by brianteeman** [comment]
And are they shown after you make the change from the message

On 13 April 2016 at 07:01, Franz Wohlkönig <notifications@github.com> wrote:

> I have tested this item [image: :white_check_mark:] successfully on
> 2516589
> <https://github.com/joomla/joomla-cms/commit/25165893ec8ef1e474d06feb5da501432afd9c3e>
>
> Tested to create sass/scss-Files on Beez3, Hathor, Isis and Protostar. Got
> now Message why Files aren't shown.
> ------------------------------
> This comment was created with the J!Tracker Application
> <https://github.com/joomla/jissues> at issues.joomla.org/joomla-cms/9871
> <https://issues.joomla.org/tracker/joomla-cms/9871>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9871#issuecomment-209241631>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by franz-wohlkoenig** [comment]
I thought its enough to know why the Files don't show. I don't know so i have to search how to "append the new File-Extensions in the List of supported Formats at the Template-Options".

**Comment by brianteeman** [comment]
On the main template manager screen select options from the toolbar in the top right.
Then add sass and scss to the list of Valid Source Formats
eg
txt,less,ini,xml,js,php,css,sass,scss

**Comment by franz-wohlkoenig** [comment]
Thanks @brianteeman Yes, they are now shown.

**Comment by brianteeman** [comment]
Thanks for testing

On 13 April 2016 at 09:21, Franz Wohlkönig <notifications@github.com> wrote:

> Thanks @brianteeman <https://github.com/brianteeman> Yes, they are now
> shown.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9871#issuecomment-209299489>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by andrepereiradasilva** [comment]
so what's needed to test now ?

**Comment by brianteeman** [comment]
@andrepereiradasilva as before but if after applying the patch you have not updated the list of allowed extensions in the options you will still be able to create a file but you cant see it and so a message will be displayed

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 25165893ec8ef1e474d06feb5da501432afd9c3e<br /><br />I think what was left to test is the message when you don't have the filetype in the template config options.

If so, tested successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by andrepereiradasilva** [comment]
ok, just tested. :)

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9871">issues.joomla.org/joomla-cms/9871</a>.</sub>

**Comment by rdeutz** [comment]
@brianteeman cloud you please check the merge conflicts, thanks

File: administrator/components/com_templates/models/template.php [hunk 2]
```diff
@@ -1386,4 +1390,29 @@ public function extractArchive($file)
...
 	}
+
+	/**
+	 * Check if the extension is allowed and will be shown in the template manager
+	 *
+	 * @param   string  $ext  The extension to check if it is allowed
+	 *
+	 * @return  boolean  true if the extension is allowed false otherwise
+	 *
+	 * @since   3.5.3
```
Comment (zero-24) [inline administrator/components/com_templates/models/template.php]: @brianteeman can you fix than this since tag to 3.6 too?

**Comment by wojsmol** [comment]
@brianteeman Plese remove conficts markers

**Comment by brianteeman** [comment]
Closed see #10186

**Pull request closed without merging**